jcl/src/java.base/share/classes/java/lang/Class.java

```diff
@@ -1549,9 +1560,89 @@ Method getMethodHelper(
 			}
 		}
 	}
+	
+	/* The more specific method is retrieved via the top superclass rather than the direct interface
+	 * of this Class only if the method is also declared by another interface(retrieved via the top superclass)
+	 * which is not the same or super-interface of the direct interface of this Class.
+	 */
+	if (!forDeclaredMethod && (resultFromSuperClass != null)) {
+		candidateFromTopSuperClass = getMostSpecificMethodFromSuperClass(resultFromSuperClass, name, strSig, parameterTypes);
+		
+		Class <?> declaringClassOfBestCandidate = bestCandidate.getDeclaringClass();
+		Class <?> declaringClassFromTopSuperClass = candidateFromTopSuperClass.getDeclaringClass();
+		if (declaringClassOfBestCandidate.isInterface()
+		&& declaringClassFromTopSuperClass.isInterface()
+		&& !declaringClassFromTopSuperClass.isAssignableFrom(declaringClassOfBestCandidate)
+		&& bestCandidate.getReturnType().isAssignableFrom(candidateFromTopSuperClass.getReturnType())
+		) {
+			bestCandidate = candidateFromTopSuperClass;
+		}
+	}
+	
 	return cacheMethod(bestCandidate);
 }
 
+/**
+ * Helper method retrieves the specified method implemented by the superclass from the top to the bottom
+ */
+private Method getMethodImplementedBySuperClass(Class<?> currentClass, String name, String strSig, Class<?>... parameterTypes) {
```

Comment:
The reason I avoided recursion to retrieve the super class implementing specified interfaces is because it might incur overhead of method calls but I agree iteration here might make things a little bit complicated.

@pshipton,  any input as to recursion vs iteration in this case?

Reply:
I think the recursion for this is fine, and preferred since it is much simpler.

Reply:
I replaced with the iterative code with the recursion as suggested above and it works fine with the failing tests:
```
1. Interface I is implemented by D:
interface I { void m(); }
interface J { void m(); }
interface K extends I, J {  }
abstract class C {  }
abstract class D extends C implements I {  } <-----------
abstract class E extends D implements J, K {  }

.../getMethodTest# ../jdk11_openj9_recursive/bin/java  -cp .  getMethodTest
testMethod.classMethod = public abstract void I.m()


2. Interface I is implemented by D and Interface L is implemented by C:
interface I { void m(); }
interface J { void m(); }
interface L { void m(); }  <-------
interface K extends I, J {  }
abstract class C implements L {  } <------
abstract class D extends C implements I {  }  <------
abstract class E extends D implements J, K {  }

.../getMethodTest2# ../jdk11_openj9_recursive/bin/java  -cp .  getMethodTest2
testMethod.classMethod = public abstract void L.m()


3.The failing test cases passed
.../jdk11_PublicMethods# ../jdk11_openj9_recursive/bin/java   -cp .  PublicMethodsTest
.../jdk11_PublicMethods# (all tests passed without exception)
```
Will launch personal builds to double-check.